src/core/ObservableQuery.ts

```diff
@@ -252,7 +252,8 @@ export class ObservableQuery<
     if (
       // These fetch policies should never deliver data from the cache, unless
       // redelivering a previously delivered result.
-      skipCacheDataFor(fetchPolicy) ||
+      (skipCacheDataFor(fetchPolicy) &&
+        !(fetchPolicy === "network-only" && !this.waitForOwnResult)) ||
```

Comment:
I always find the `!(...)` harder to reason about since you have to understand the stuff inside the parens, then negate it. Instead, we can apply [De Morgan's Law](https://en.wikipedia.org/wiki/De_Morgan%27s_laws) which states that "not (A and B)" is equaivalent of "(not A) or (not B)" (essentially distributing the `!` to each clause and changing the `&&` to an `||`.

I find that a bit easier to reason about:

```suggestion
        (fetchPolicy !== "network-only" || this.waitForOwnResult)) ||
```

Reply:
Ok now to dissect this further. For me to understand this better, it helps for me to do a bit of refactoring to remove the function calls and inline the checks here. Let's first start with [inline method](https://refactoring.guru/inline-method) to pull in the body of `skipCacheData` in place of its check.

```ts
((fetchPolicy === "no-cache" ||
  fetchPolicy === "network-only" ||
  fetchPolicy === "standby") &&
  (fetchPolicy !== "network-only" || this.waitForOwnResult))
```

This now makes it possible to see the combinations of `fetchPolicy` checks here.

Digging further, we know that the 2nd clause won't be executed unless the first is truthy. Using this, we can deduce that `fetchPolicy !== 'network-only'` is the same as `fetchPolicy === 'no-cache' || fetchPolicy === 'standby'` since these are the only two fetch policies left over from the first check.

```ts
((fetchPolicy === "no-cache" ||
  fetchPolicy === "network-only" ||
  fetchPolicy === "standby") &&
  (fetchPolicy === "no-cache" || fetchPolicy === 'standby' || this.waitForOwnResult))
```

We now begin to really see some repetition here, notably that when `fetchPolicy === 'no-cache' || fetchPolicy === 'standby'` is true, the whole thing will always be true. 

Let's temporarily refactor out this check to its own variable to make the conditional a bit more obvious.

```ts
const absoluteNoCache =
  fetchPolicy === "no-cache" || fetchPolicy === "standby";

// ...

((absoluteNoCache || fetchPolicy === "network-only") &&
  (absoluteNoCache || this.waitForOwnResult))
```

This makes it obvious that the condition will pass whenever `absoluteNoCache` is `true`. We should be able to pull `absoluteNoCache` out into its own clause and "or" it with the rest of the logic.

```ts
(absoluteNoCache || ((fetchPolicy === "network-only") && (this.waitForOwnResult)))
```

which is the same as the following when removing unnecessary parens

```ts
(absoluteNoCache || (fetchPolicy === "network-only" && this.waitForOwnResult))
```

Now that we've got the common bits together, let's inline the variable back into the conditional

```ts
(fetchPolicy === "no-cache" || fetchPolicy === "standby" || (fetchPolicy === "network-only" && this.waitForOwnResult))
```

Here is the total of the `if` altogether:

```ts
    if (
      // These fetch policies should never deliver data from the cache, unless
      // redelivering a previously delivered result.
      fetchPolicy === "no-cache" ||
      fetchPolicy === "standby" ||
      (fetchPolicy === "network-only" && this.waitForOwnResult) ||
      // If this.options.query has @client(always: true) fields, we cannot
      // trust diff.result, since it was read from the cache without running
      // local resolvers (and it's too late to run resolvers now, since we must
      // return a result synchronously).
      this.queryManager.getDocumentInfo(this.query).hasForcedResolvers
    ) {
```

Unfortunately I have no tests to check my work, but hopefully stepping through this makes sense.

I don't have a great recommendation other than the fix you have here. I'd just suggest doing the above refactoring and if possible, add a test, and I'll be happy with this change.